### PR TITLE
Function to check if additional repos are present after the upgrade of admin node

### PR DESCRIPTION
**Pull request by jsuchome**
Requires https://github.com/crowbar/crowbar-core/pull/188 and https://github.com/crowbar/crowbar-ceph/pull/34

Function like this needs to be called from upgrade wizard

### CODE REVIEW

**Comment by aplanas** [comment]
lgtm so far, from where is going to be called?

**Comment by jsuchome** [comment]
Not sure. I expect from upgrade wizard (which does not exist yet), in a loop, until it returns true.

**Comment by aplanas** [comment]
Are the i18n string also there? Actually I do not find them

**Comment by jsuchome** [comment]
The one for pacemaker should be there - I haven't checked, but the code is copied from crowbar-ha.

File: crowbar_framework/lib/openstack/upgrade.rb [hunk 1]
```diff
@@ -45,5 +45,26 @@ def self.unset_db_synced
...
     end
+
+    # Check if the additional repositories (HAE, SES) are present
+    # after the upgrade of the admin node
+    def self.check_additional_repositories_presence
+
+      # check if HA repos are needed
+      pacemaker_proposal = Proposal.where(barclamp: "pacemaker").first
+      unless (pacemaker_proposal.nil? || Crowbar::Repository.provided_and_enabled? ("ha"))
+        validation_error I18n.t("barclamp.pacemaker.validation.hae_repo")
```
Comment (vuntz) [inline crowbar_framework/lib/openstack/upgrade.rb]: Is `validation_error` available from here?

Comment (jsuchome) [inline crowbar_framework/lib/openstack/upgrade.rb]: It is not.

File: crowbar_framework/lib/openstack/upgrade.rb [hunk 2]
```diff
@@ -45,5 +45,21 @@ def self.unset_db_synced
         node.save if save_it
       end
     end
+
+    # Check if the additional repositories (HAE, SES) are present
+    # after the upgrade of the admin node
+    def self.check_additional_repositories_presence
+      # check if HA repos are needed
+      pacemaker_proposal = Proposal.where(barclamp: "pacemaker").first
+      unless pacemaker_proposal.nil? || (Crowbar::Repository.provided_and_enabled? "ha")
```
Comment (jsuchome) [inline crowbar_framework/lib/openstack/upgrade.rb]: Looks like we will need to split the check for "provided" and "enabled"

**Comment by vuntz** [comment]
+1. We could possibly move the checks to crowbar-ha and crowbar-ceph later on -- but we'll figure this out when we know how this will be called.

**Comment by aplanas** [comment]
+1

**Comment by jsuchome** [comment]
I've added dependency on https://github.com/crowbar/crowbar-ceph/pull/34

File: crowbar_framework/lib/openstack/upgrade.rb [hunk 3]
```diff
@@ -45,5 +45,36 @@ def self.unset_db_synced
...
+        if (repo.config["features"] || []).include? "ha"
+          Rails.logger.info("enabling repository #{repo.id}")
+          @provisioner_service.enable_repository(repo.platform, repo.arch, repo.id)
+        end
+      end
+
+      # check if Storage repos are neededa
+      ceph_proposal = Proposal.where(barclamp: "ceph").first
+      unless ceph_proposal.nil?
+        unless Crowbar::Repository.provided_and_enabled? "ceph"
```
Comment (vuntz) [inline crowbar_framework/lib/openstack/upgrade.rb]: Hrm, should be `provided?` I think.

**Comment by jsuchome** [comment]
(rebased)

File: crowbar_framework/lib/openstack/upgrade.rb [hunk 4]
```diff
@@ -45,5 +45,36 @@ def self.unset_db_synced
         node.save if save_it
       end
     end
+
+    # Check if the additional repositories (HAE, SES) are present
+    # after the upgrade of the admin node
+    def self.check_additional_repositories_presence
+      @provisioner_service = ProvisionerService.new Rails.logger
```
Comment (rsalevsky) [inline crowbar_framework/lib/openstack/upgrade.rb]: The logger should be handed over by the function which calls the lib.

**Comment by jsuchome** [comment]
@vuntz could you please merge this, it's out of date again...

File: crowbar_framework/lib/openstack/upgrade.rb [hunk 5]
```diff
@@ -45,5 +45,36 @@ def self.unset_db_synced
...
+    def self.check_additional_repositories_presence(thelogger)
+      @provisioner_service = ProvisionerService.new thelogger
+      # check if HA repos are needed
+      pacemaker_proposal = Proposal.where(barclamp: "pacemaker").first
+      unless pacemaker_proposal.nil?
+        unless Crowbar::Repository.provided? "ha"
+          raise I18n.t("barclamp.pacemaker.validation.hae_repo")
+        end
+        # enable disabled repos now
+        if (repo.config["features"] || []).include? "ha"
```
Comment (vuntz) [inline crowbar_framework/lib/openstack/upgrade.rb]: Sorry, before merging, one last question: where is `repo` defined? Or don't you want to iterate over all repos?

Comment (jsuchome) [inline crowbar_framework/lib/openstack/upgrade.rb]: Ouch, indeed. I must have lost it somewhere...

**Comment by jsuchome** [comment]
Hopefully fixed last part, but it's untested now...

**Comment by vuntz** [comment]
+1

**Comment by nkrinner** [comment]
+1

**Comment by nkrinner** [comment]
Ran a rudimentary but successful test. Seems to work fine.

**Pull request merged**